Sources/Apodini/RHI/RHIInterfaceExporter.swift

```diff
@@ -0,0 +1,100 @@
+        private let imp: (_ endpointParamId: UUID) -> Any?
+        
+        init<H: InvocableHandler>(endpoint: Endpoint<H>, collectedParameters: [CollectedParameter<H>]) {
+            self.imp = { endpointParamId -> Any? in
+                collectedParameters.first { collectedParam in
+                    (endpoint.handler[keyPath: collectedParam.handlerKeyPath] as? AnyParameterID)?.value == endpointParamId
+                }?.value
+            }
+        }
+        
+        func getValueOfCollectedParameter(for endpointParameter: AnyEndpointParameter) -> Any? {
+            imp(endpointParameter.id)
+        }
+    }
+    
+    internal private(set) static var shared: RHIInterfaceExporter?
+    
+    let app: Apodini.Application
+    private var endpointsById: [AnyHandlerIdentifier: AnyEndpoint] = [:]
+    
+    
+    required init(_ app: Apodini.Application) {
+        self.app = app
+        // NOTE: if this precondition fails while running tests, chances are you have to call `RHIInterfaceExporter.resetSingleton` in your -tearDown method
+        precondition(Self.shared == nil, "-[\(Self.self) \(#function)] cannot be called multiple times")
+        Self.shared = self
+    }
+    
+    
+    func export<H: Handler>(_ endpoint: Endpoint<H>) {
+        endpointsById[endpoint.identifier] = endpoint
+    }
+    
+    
+    func finishedExporting(_ webService: WebServiceModel) {
+        // This is where we will expose the remote-invocation endpoint
+//        app.vapor.app.post(["__apodini", "invoke"]) { request -> String in
+//            String(reflecting: request)
+//        }
+    }
+    
+    
+    func getEndpoint<H: IdentifiableHandler>(withIdentifier identifier: H.HandlerIdentifier, ofType _: H.Type) -> Endpoint<H>? {
+        endpointsById[identifier] as? Endpoint<H>
+    }
+    
+    
+    func retrieveParameter<Type: Codable>(_ endpointParameter: EndpointParameter<Type>, for request: ExporterRequest) throws -> Type?? {
+        guard let value: Any = request.getValueOfCollectedParameter(for: endpointParameter) else {
+            return Optional<Type?>.none // this should be a "top-level" nil value (ie `.none` instead of `.some(.none)`)
+        }
+        
+        if endpointParameter.nilIsValidValue {
+            if let value: Type? = dynamicCast(value, to: Type?.self) {
+                return .some(value)
+            }
+        } else {
+            if let value = value as? Type {
+                return .some(.some(value))
+            }
+        }
+        throw makeApodiniError("Unable to cast parameter value (of type '\(type(of: value))' to expected type '\(Type.self)'")
+    }
+}
+
+
+extension RHIInterfaceExporter {
+    // Used by the tests to get a new object for every test case.
+    // Ideally this function would be wrapped in some `#if TEST` condition, but that doesn't seem to be a thing
+    internal static func resetSingleton() {
+        Self.shared = nil
+    }
+}
```

Comment:
The extension needs to be in the same file as the class definition because `shared`'s setter is private. If we move the extension somewhere else, the setter would need to be internal. The reason why this function exists (instead of just making the setter internal) is so that the singleton can, from outside the class, only be reset to `nil` but never set to a nonnil value (which would be possible if the setter isn't private anymore).

Reply:
Thanks for clearing this up 👍